integration_tests/src/fixture_generator/parser.rs

```diff
@@ -0,0 +1,153 @@
+}
+
+/// This tool allows a directory to be created with specified and random
+/// contents, capturing the directory state and epoch-to-epoch delta in
+/// an output file for use in debugging and as test fixtures.
+#[derive(Parser, Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[clap(setting = AppSettings::DeriveDisplayOrder)]
+pub struct Args {
+    /// Users and their associated update events.
+    /// A username is expected, followed by a colon and a list of epochs OR
+    /// (epoch, value). Usernames are expected to be utf-8 strings, which will
+    /// be internally interpreted as bytes.
+    /// The following are valid examples of user arguments:
+    ///   --user "username: 1, 3, (5, 'xyz')"
+    ///   --user="username: [(1,'abc'), 2]"
+    ///   -u "some username: 1"
+    #[clap(
+        long = "user",
+        short = 'u',
+        multiple_occurrences = true,
+        parse(try_from_str = parse_user_history),
+    )]
+    pub users: Vec<User>,
+
+    /// Number of epochs to advance the tree by
+    /// e.g. a value of 3 will perform 3 publishes on an empty directory.
+    #[clap(long = "epochs", short = 'e')]
+    pub epochs: u32,
+
+    /// Maximum number of updates **per epoch** the tool should perform.
+    /// Note that all user events explicitly passed for an epoch will be
+    /// included even if the number exceeds this value.
+    #[clap(long = "max_updates", default_value = "10")]
+    pub max_updates: u32,
+
+    /// Minimum number of updates **per epoch** the tool should perform.
+    /// The tool will generate random labels and values to include in an epoch
+    /// if the user events explicitly passed for an epoch are not sufficients.
+    #[clap(long = "min_updates", default_value = "0")]
+    pub min_updates: u32,
+
+    /// Epochs where the state of the directory should be captured in the output
+    /// e.g. the value 3 will output all db records after epoch 3 is performed.
+    /// Multiple values should be separated with ","
+    #[clap(
+        long = "capture_states",
+        short = 's',
+        multiple_values = true,
+        use_value_delimiter = true,
+        require_value_delimiter = true
+    )]
+    pub capture_states: Option<Vec<u32>>,
+
+    /// Epochs where the updates required to bring the directory to the epoch
+    /// should be captured in the output.
+    /// e.g. the value 3 will output all updates that were performed to advance
+    /// the directory from epoch 2 to 3.
+    /// Multiple values should be separated with ","
+    #[clap(
+        long = "capture_deltas",
+        short = 'd',
+        multiple_values = true,
+        use_value_delimiter = true,
+        require_value_delimiter = true
+    )]
+    pub capture_deltas: Option<Vec<u32>>,
+
+    /// Name of output file.
+    /// If omitted, output will be printed to stdout.
+    #[clap(long = "out", short = 'o')]
+    pub out: Option<String>,
+
+    /// Stops tool from generating random updates in publishes.
```

Comment:
If generated updates are used, do we ignore user-provided values? Could we just go with if no value is provided, we will generate a random one?

Reply:
Clarification: an update consists of both a _label_ and a _value_.

So by default random updates are used in conjunction with user-provided updates. In an epoch, the tool will first add all user-provided updates, then pick a value between `--min_updates` and `--max_updates` as the number of updates required. If the user-provided updates are less than the required, the tool will randomly generate the rest.

This is not to be confused with if users specify an update epoch, but omit a value. In that case a random value is always generated.